libp2p/pubsub/pubsub.py

```diff
@@ -417,10 +438,14 @@ def handle_subscription(
         logger.debug("unsubscribing from topic %s", topic_id)
 
         # Return if we already unsubscribed from the topic
-        if topic_id not in self.my_topics:
+        if topic_id not in self.topic_ids:
             return
-        # Remove topic_id from map if present
-        del self.my_topics[topic_id]
+        # Remove topic_id from the maps before yielding
+        send_channel = self.subscribed_topics_send[topic_id]
+        del self.subscribed_topics_send[topic_id]
+        del self.subscribed_topics_receive[topic_id]
```

Comment:
I think it would be ideal for the `SubscriptionAPI` to be in control of doing all of this cleanup in the `__aexit__`.  The way I've done this elsewhere is to have the subscription API work something like this.

```python
class Subscription:
    def __init__(self, unsubscribe_fn, ...)
        self._unsubscribe_fn = unsubscribe_fn

    def unsubscribe(self):
        self._unsubscribe_fn()

    def __aexit__(self, ...):
        self.unsubscribe()
```

Then in your function here where you create the subscription you do this.

```python
def subscribe(self, ...) -> Subscription:
    subscription = Subscription(
        unsubscribe_fn=functools.partial(self,unsubscribe, topic_id)
        ...
    )
    ...
    return subscription
```

And this inturn allows for usage like this.

```python
async with pubsub.subscribe(topic_id) as subscription:
    ...  # active here
# will be cleaned up and removed once the context exits
```

Reply:
This is really nice! I will change it in the next commits.